### ISSUE TITLE
Data.copy_compatible_to does not add feature dim

### ISSUE THREAD

**Issue by Zettelkasten** [rwth-i6/returnn#428]
I have some tensor without a feature-dim, e.g. `[T]`, and want to copy it compatible to one with a feature dim, e.g. `[T,F]`, like in this test case here (`test_Data_copy_compatible_to_add_feature_dim`):

```python3
common_data = Data(name='energy', shape=(None, 3), batch_dim_axis=None)  # [T,F]
d1 = Data(name='mask', shape=(None,), batch_dim_axis=None, time_dim_axis=0, feature_dim_axis=None)  # [T]
d2 = d1.copy_compatible_to(common_data)
```

That however fails:
```
  File "/u/petrick/software/returnn/official/returnn/tf/util/data.py", line 1278, in copy_compatible_to
    line: assert data.feature_dim_axis is not None and v.feature_dim_axis is not None
    locals:
      data = <local> Data(name='energy', shape=(None, 3), batch_dim_axis=None, time_dim_axis=None, batch_shape_meta=[?,F|3])
      data.feature_dim_axis = <local> 1
      v = <local> Data(name='mask', shape=(1, None), batch_dim_axis=None, time_dim_axis=1, feature_dim_axis=None, batch_shape_meta=[1,T|?])
      v.feature_dim_axis = <local> None
```

I tried figuring out what went wrong in `Data.copy_compatible_to`, but I can't figure it out.
It would be great if somebody could take a look!

I pushed three test cases to 03416cdc3f0fe1b34bd20506ea0d88456b69528a. Note that `test_Data_copy_compatible_to_add_time_dim` already passes, this issue only occurs with the feature dim but not when adding a time dim.

**Comment by albertz** [comment]
From briefly looking at it, I think `copy_compatible_to` gets confused about the spatial dim, which is unknown (without dim tag / seq len).

**Comment by albertz** [comment]
We miss the logic (workaround / heuristic) which we have in `find_matching_dim_map`. I.e. to match an unknown time-dim if it's unique.

**Comment by Zettelkasten** [comment]
I tried fiddling around with this for a while now, but ran into a few issues: I think the problem is that I don't quite know how `copy_compatible` exactly worked before, and how exactly it should behave in the future.

I tried to replace the call to `DimensionTag.get_all_dimension_tags([start, data], dict(allow_same_feature_dim=True))` with `find_matching_dim_map`, but that function cannot handle all cases properly as `get_all_dimension_tags` did before: When more than one axis does not have a `DimesionTag`, this fails.

For example look at
`Data(name='p1_output', shape=(5, 5, 3), batch_dim_axis=None, time_dim_axis=None, batch_shape_meta=[5,5,F|3]),
Data(name='p2_output', shape=(5, 5, 3), batch_dim_axis=None, time_dim_axis=None, batch_shape_meta=[5,5,F|1])`,
which are used in [`test_CombineLayer_broadcast_multiple`](https://github.com/rwth-i6/returnn/blob/master/tests/test_TFNetworkLayer.py#L512).

In this case of course, we want to keep the order of the axes.
And in general, this sound like a good idea to do when something is unclear, but I don't know what the exact rules should be.

Then, there are also cases like this, e.g. in [`test_nd_indices_scatter_nd_time_major`](https://github.com/rwth-i6/returnn/blob/master/tests/test_TFUtil.py#L1370), which tries to copy
`Data(name='v', shape=(2, 5), time_dim_axis=None, batch_shape_meta=[B,2,F|5])` compatible to `Data(name='x', shape=(None, 5), batch_shape_meta=[B,T|"time:var:'Const_1:0':x",F|5])`.
Should that even be allowed?

Ideally, I think it would be great if this was all handled by `find_matching_dim_map`. If we have a proper implementation of this, then the logic in `copy_compatible_to` actually is quite straight forward and easy to understand:

```python3
# data = "target", v = "to adapt to target"
is_equal_opts = dict(
  allow_same_feature_dim=True, allow_same_spatial_dim=True, treat_feature_as_spatial=True, ignore_feature_dim=True)
mapped_axes = data.find_matching_dim_map(start, list(range(v.batch_ndim)), is_equal_opts)  # maps start -> data
assert len(mapped_axes) == start.batch_ndim

for target_axis in range(data.batch_ndim):
  new_v_axis = min(target_axis, v.batch_ndim)
  if target_axis not in mapped_axes.values():
    # Dim in data, but not in v
    v = v.copy_add_dim_by_tag(data.get_dim_tag(target_axis), axis=new_v_axis)
    # Keep mapped_axes consistent
    mapped_axes = {v_ax + (1 if v_ax >= new_v_axis else 0): trg_ax for v_ax, trg_ax in mapped_axes.items()}
    mapped_axes[new_v_axis] = target_axis
  else:
    # Dim exists in both data and in v. Maybe order is wrong.
    matching_v_axes = [v_ax for v_ax, trg_ax in mapped_axes.items() if trg_ax == target_axis]
    assert len(matching_v_axes) == 1
    matching_v_axis = matching_v_axes[0]
    if target_axis != matching_v_axis:
      # Order was wrong
      v = v.copy_move_axis(matching_v_axis, new_v_axis, keep_other_axes_intact=True)  # `keep_other_axes_intact` does not exist yet
      # Keep mapped_axes consistent
      mapped_axes[matching_v_axis], mapped_axes[new_v_axis] = mapped_axes[new_v_axis], mapped_axes[matching_v_axis]

assert v.batch_ndim == data.batch_ndim
assert all(mapped_axes[ax] == ax for ax in range(v.batch_ndim))
```

This code of course is lacking some smaller details the old `copy_compatible_to` implementation had (like `unbroadcast`), but I think it is much easier to maintain and clearer in what it does than the current implementation.

So my main questions are:
- Does the sketch of my new implementation do what `copy_compatible_to` should do? (Just idea wise, of course it has to pass all tests)
- How should `find_matching_dim_map` work to handle cases like in [`test_CombineLayer_broadcast_multiple`](https://github.com/rwth-i6/returnn/blob/master/tests/test_TFNetworkLayer.py#L512)? Should something like [`test_nd_indices_scatter_nd_time_major`](https://github.com/rwth-i6/returnn/blob/master/tests/test_TFUtil.py#L1370) even be allowed?
- Minor details: What should the `feature_dim_axis` / `time_dim_axis` of the output of `copy_compatible_to` be set to? Always equal to how they are set in `data`, or how they were set in `v` (but with moved axes?)

Also, I think we should add the answer to most of these questions to the documentation.

**Comment by albertz** [comment]
I'm quite busy recently, so just some comments.

> I think the problem is that I don't quite know how copy_compatible exactly worked before, and how exactly it should behave in the future.

I guess the intent is clear, right? "compatible" means that it ends up with same number of dimensions, and all broadcast-able operations (e.g. simply `+` etc) work correctly, i.e. the dimensions are in the correct order. It should add broadcast dimensions for any missing dimensions.

It is used by e.g. `CombineLayer`, which is exactly intended to combine multiple sources (e.g. via `add`, `mul`, whatever, or custom eval str).

> I tried to replace the call to `get_all_dimension_tags` with `find_matching_dim_map`

Yea, I also briefly looked at this, and I came to the same conclusion, that `find_matching_dim_map` is unfortunately not generic enough. But it could maybe still partly share same code, or at least same logic / behavior. But also, let's not over-engineer things, and make things more complicated than they need to be. The priority should be that the code should also be easy to read and understand (and admittedly, it's already not so optimal in this respect).

> Data(name='p2_output', shape=(5, 5, 3), ... batch_shape_meta=[5,5,F|1])

I guess you have a typo there, right?

> In this case of course, we want to keep the order of the axes.
> And in general, this sound like a good idea to do when something is unclear, but I don't know what the exact rules should be.

Yea, that's the tricky part, when sth is not clear. The matching of axes is not perfect. First of all, because dynamic axes can be unknown (which is a problem we will hopefully clean up at some point). Then, already existing broadcast dimensions can arbitrarily match. I also don't really know about the optimal solution. The current logic also mostly has grown historically. But luckily we have all the test cases.

> test_nd_indices_scatter_nd_time_major ...

I don't exactly remember anymore why this was done...
Sometimes test cases itself can also be buggy, and we just did not notice before...

**Comment by Zettelkasten** [comment]
Thanks for your comments, @albertz!

I think I sorted out all issues I have stumbled apon.
In #429 I added a commit cleaning up and essentially rewriting `Data.copy_compatible_to`, which now uses the logic of `Data.find_matching_dim_map`, which I extended with an additional check to match broadcasted axes.
`Data.copy_compatible_to` is now also using a newly added `different_static_matches` parameter in `DimensionTag.is_equal`.

Note however that the new logic is a bit more restrictive than the old one.
Previously, e.g. the aforementioned `test_nd_indices_scatter_nd_time_major` test trying to match a (non 1) static dimension to a dynamic axis is no longer allowed and will fail.
I am skipping that specific test case for now.
If we really want to, we can re-add that functionality. But I would argue that those shapes are actually not compatible and it is a better idea to just fail - it is very easy to get something wrong if we match too many axes wildly with each other.
Also, this is the only test case that relies on this behavior.
Maybe however some configs use this in an `EvalLayer`? (Note however that we still match it if the static dim is `1` to support broadcasting, as well as any dim as long as both are static).

> > Data(name='p2_output', shape=(5, 5, 3), ... batch_shape_meta=[5,5,F|1])
> 
> I guess you have a typo there, right?

Ah yes, my bad. It's supposed to have shape `(5, 5, 3)`.
Because my commit now uses `different_static_matches=True`, those axes are matched now.
